### DISCUSSION TITLE
Requested lower limits aren't applied.

### DISCUSSION THREAD

**Discussion by Kangz** [gpuweb/gpuweb#2528] [Q&A] [answered]
@onsummer asked:

> Question, when I requestDevice on Edge Canary latest, I pass a GPUDeviceDescriptor obj:
> 
> ```js
> const deviceLimitTextureDim2D = await adapter.requestDevice({
>   requiredLimits: {
>     maxTextureDimension2D: 2048
>   }
> })
> ```
> 
> Then request successful, and I log `deviceLimitTextureDim2D.limits`, but `deviceLimitTextureDim2D.limits.maxTextureDimension2D` is still 8192, not expect 2048.
> 
> It is the same result in deno 1.18, why?

**Comment by wewindy** [comment]
in deno 1.18:

![image](https://user-images.githubusercontent.com/26168955/150639024-b23526cd-4974-40cf-bf8e-879deef9a10f.png)

in edge canary latest:

![image](https://user-images.githubusercontent.com/26168955/150639061-bb5bd024-2d78-4757-9f24-256ace644dd2.png)

**Comment by wewindy** [comment, marked as answer]
OK, I read the webgpu specification again, Edge is right, my fault.

**Reply by Kangz** [reply]
Can you quote the relevant spec text here? I wasn't able to find it, and it will help other people that hit a similar issue know what to do.

**Reply by wewindy** [reply]
https://www.w3.org/TR/webgpu/#devices

> When a new device device is created from adapter adapter with GPUDeviceDescriptor descriptor:
> - Set device.[[adapter]] to adapter.
> - Set device.[[features]] to the set of values in descriptor.requiredFeatures.
> - Let device.[[limits]] be a supported limits object with the default values. For each (key, value) pair in descriptor.requiredLimits, set the member corresponding to key in device.[[limits]] to the better value of value or the default value in supported limits.

The creation process of GPUDevice has been described in detail above
